File: yaeui/yaeAudioUnitRenderer.cpp
// -*- Mode: c++; tab-width: 8; c-basic-offset: 2; indent-tabs-mode: nil -*-
// NOTE: the first line of this file sets up source code indentation rules
// for Emacs; it is also a hint to anyone modifying this file.

// Created      : Sat Jul  8 13:10:51 MDT 2017
// Copyright    : Pavel Koshevoy
// License      : MIT -- http://www.opensource.org/licenses/mit-license.php

// aeyae:
#include "yae/video/yae_audio_renderer_input.h"

// standard:
#include <iostream>
#include <stdexcept>

YAE_DISABLE_DEPRECATION_WARNINGS

// boost:
#ifndef Q_MOC_RUN
#include <boost/atomic.hpp>
#include <boost/thread.hpp>
#endif

YAE_ENABLE_DEPRECATION_WARNINGS

// yaeui:
#include "yaeAudioUnitRenderer.h"


namespace yae
{
  //----------------------------------------------------------------
  // AudioUnitRenderer::TPrivate
  //
  class AudioUnitRenderer::TPrivate
  {
  public:
    TPrivate(SharedClock & sharedClock);
    ~TPrivate();

    void match(const AudioTraits & source,
               AudioTraits & output) const;

    bool open(IReader * reader);
    void stop();
    void close();

  private:
    static bool pull_cb(void * context, // this
                        void * data,
                        unsigned long samples_to_read,
                        int channel_count,
                        bool planar);

    static void stop_cb(void * context); // this

    // opaque:
    void * yae_au_ctx_;

  public:
    // protect against concurrent access:
    mutable boost::timed_mutex mutex_;

    // audio source:
    AudioRendererInput input_;
    boost::atomic<bool> stop_;
  };

  //----------------------------------------------------------------
  // AudioUnitRenderer::TPrivate::TPrivate
  //
  AudioUnitRenderer::TPrivate::TPrivate(SharedClock & sharedClock):
    input_(sharedClock),
    stop_(true)
  {
    yae_au_ctx_ = yae_au_ctx_create(this, &pull_cb, &stop_cb);
  }

  //----------------------------------------------------------------
  // AudioUnitRenderer::TPrivate::~TPrivate
  //
  AudioUnitRenderer::TPrivate::~TPrivate()
  {
    close();
    yae_au_ctx_destroy(&yae_au_ctx_);
  }

  //----------------------------------------------------------------
  // AudioUnitRenderer::TPrivate::match
  //
  void
  AudioUnitRenderer::TPrivate::match(const AudioTraits & srcAtts,
                                     AudioTraits & outAtts) const
  {
    if (&outAtts != &srcAtts)
    {
      outAtts = srcAtts;
    }

    outAtts.sample_format_ = AV_SAMPLE_FMT_FLT;
  }

  //----------------------------------------------------------------
  // AudioUnitRenderer::TPrivate::open
  //
  bool
  AudioUnitRenderer::TPrivate::open(IReader * reader)
  {
    stop();

    boost::unique_lock<boost::timed_mutex> lock(mutex_);
#ifndef NDEBUG
    yae_debug << "AudioUnitRenderer::TPrivate::open " << reader;
#endif

    AudioTraits atts;
    if (input_.open(reader) &&
        input_.reader_->getAudioTraitsOverride(atts))
    {
      if (atts.is_invalid_format())
      {
        return false;
      }

      int sample_rate = int(atts.sample_rate_);
      int num_channels = atts.ch_layout_.nb_channels;
      if (yae_au_ctx_open_stream(yae_au_ctx_, sample_rate, num_channels))
      {
        // FIXME: is this a safe assumption?
        input_.outputLatency_ = 16e-3;
        stop_ = false;
        return true;
      }
    }

    stop_ = true;
    return false;
  }

  //----------------------------------------------------------------
  // AudioUnitRenderer::TPrivate::stop
  //
  void
  AudioUnitRenderer::TPrivate::stop()
  {
#ifndef NDEBUG
    yae_debug << "AudioUnitRenderer::TPrivate::stop";
#endif

    if (stop_)
    {
      return;
    }

    stop_ = true;
    input_.stop();

    boost::unique_lock<boost::timed_mutex> lock(mutex_);
    yae_au_ctx_stop(yae_au_ctx_);
  }

  //----------------------------------------------------------------
  // AudioUnitRenderer::TPrivate::close
  //
  void
  AudioUnitRenderer::TPrivate::close()
  {
    open(NULL);
  }

  //----------------------------------------------------------------
  // AudioUnitRenderer::TPrivate::pull_cb
  //
  bool
  AudioUnitRenderer::TPrivate::pull_cb(void * context,
                                       void * data,
                                       unsigned long samples_to_read,
                                       int channel_count,
                                       bool planar)
  {
    TPrivate * renderer = (TPrivate *)context;
    try
    {
      boost::unique_lock<boost::timed_mutex> lock(renderer->mutex_,
                                                  boost::defer_lock);

      while (!lock.timed_lock(boost::posix_time::milliseconds(50)))
      {
        if (renderer->stop_)
        {
          return false;
        }

        boost::this_thread::yield();
      }

      renderer->input_.getData(data,
                               samples_to_read,
                               channel_count,
                               planar);
      return true;
    }
    catch (const std::exception & e)
    {
#ifndef NDEBUG
      yae_error
        << "AudioUnitRenderer::TPrivate::callback: "
        << "abort due to exception: " << e.what();
#endif
    }
    catch (...)
    {
#ifndef NDEBUG
      yae_error
        << "AudioUnitRenderer::TPrivate::callback: "
        << "abort due to unexpected exception";
#endif
    }

    return false;
  }

  //----------------------------------------------------------------
  // AudioUnitRenderer::TPrivate::stop_cb
  //
  void
  AudioUnitRenderer::TPrivate::stop_cb(void * context)
  {
#ifndef NDEBUG
    yae_debug << "AudioUnitRenderer::TPrivate::stop_cb";
#endif
    TPrivate * renderer = (TPrivate *)context;
    renderer->stop();
  }


  //----------------------------------------------------------------
  // AudioUnitRenderer::AudioUnitRenderer
  //
  AudioUnitRenderer::AudioUnitRenderer():
    private_(new TPrivate(ISynchronous::clock_))
  {}

  //----------------------------------------------------------------
  // AudioUnitRenderer::~AudioUnitRenderer
  //
  AudioUnitRenderer::~AudioUnitRenderer()
  {
    delete private_;
  }

  //----------------------------------------------------------------
  // AudioUnitRenderer::create
  //
  AudioUnitRenderer *
  AudioUnitRenderer::create()
  {
    return new AudioUnitRenderer();
  }

  //----------------------------------------------------------------
  // AudioUnitRenderer::destroy
  //
  void
  AudioUnitRenderer::destroy()
  {
    delete this;
  }

  //----------------------------------------------------------------
  // AudioUnitRenderer::getName
  //
  const char *
  AudioUnitRenderer::getName() const
  {
    return typeid(*this).name();
  }

  //----------------------------------------------------------------
  // AudioUnitRenderer::match
  //
  void
  AudioUnitRenderer::match(const AudioTraits & source,
                           AudioTraits & output) const
  {
    return private_->match(source, output);
  }

  //----------------------------------------------------------------
  // AudioUnitRenderer::open
  //
  bool
  AudioUnitRenderer::open(IReader * reader)
  {
    return private_->open(reader);
  }

  //----------------------------------------------------------------
  // AudioUnitRenderer::stop
  //
  void
  AudioUnitRenderer::stop()
  {
    private_->stop();
  }

  //----------------------------------------------------------------
  // AudioUnitRenderer::close
  //
  void
  AudioUnitRenderer::close()
  {
    private_->close();
  }

  //----------------------------------------------------------------
  // AudioUnitRenderer::pause
  //
  void
  AudioUnitRenderer::pause(bool paused)
  {
#ifndef NDEBUG
    yae_debug << "AudioUnitRenderer::pause " << (paused ? "true" : "false");
#endif
    private_->input_.pause(paused);
  }

  //----------------------------------------------------------------
  // AudioUnitRenderer::skipToTime
  //
  void
  AudioUnitRenderer::skipToTime(const TTime & t, IReader * reader)
  {
    private_->input_.skipToTime(t, reader);
  }

  //----------------------------------------------------------------
  // AudioUnitRenderer::skipForward
  //
  void
  AudioUnitRenderer::skipForward(const TTime & dt, IReader * reader)
  {
    private_->input_.skipForward(dt, reader);
  }

  //----------------------------------------------------------------
  // AudioUnitRenderer::input
  //
  const AudioRendererInput &
  AudioUnitRenderer::input() const
  {
    return private_->input_;
  }

}

File: yaeui/yaeAudioUnitRenderer.m
// -*- Mode: c++; tab-width: 8; c-basic-offset: 2; indent-tabs-mode: nil -*-
// NOTE: the first line of this file sets up source code indentation rules
// for Emacs; it is also a hint to anyone modifying this file.

// Created      : Sat Aug  8 09:13:01 MDT 2020
// Copyright    : Pavel Koshevoy
// License      : MIT -- http://www.opensource.org/licenses/mit-license.php

// standard:
#include <assert.h>
#include <stdlib.h>

// apple:
#import <Cocoa/Cocoa.h>
#import <CoreAudio/CoreAudio.h>
#import <CoreServices/CoreServices.h>
#import <AudioUnit/AudioUnit.h>
#import <AudioToolbox/AudioToolbox.h>

// yaeui:
#include "yaeAudioUnitRenderer.h"


//----------------------------------------------------------------
// yae_au_ctx_t
//
typedef struct yae_au_ctx_t
{
  void * owner_;
  TAuCtxPull pull_;
  TAuCtxStop stop_;
  AudioUnit au_;
  AudioStreamBasicDescription sd_;
} yae_au_ctx_t;

//----------------------------------------------------------------
// yae_au_ctx_create
//
void *
yae_au_ctx_create(void * owner,
                  TAuCtxPull pull,
                  TAuCtxStop stop)
{
  yae_au_ctx_t * ctx = (yae_au_ctx_t *)malloc(sizeof(yae_au_ctx_t));
  memset(ctx, 0, sizeof(*ctx));
  ctx->owner_ = owner;
  ctx->pull_ = pull;
  ctx->stop_ = stop;
  return ctx;
}

//----------------------------------------------------------------
// yae_au_ctx_destroy
//
void
yae_au_ctx_destroy(void ** context)
{
  yae_au_ctx_t * ctx = (yae_au_ctx_t *)(*context);
  if (ctx)
  {
    yae_au_ctx_stop(ctx);
    free(ctx);
    *context = NULL;
  }
}

//----------------------------------------------------------------
// yae_au_ctx_stop
//
void
yae_au_ctx_stop(void * context)
{
  yae_au_ctx_t * ctx = (yae_au_ctx_t *)context;
  if (ctx->au_)
  {
    AudioOutputUnitStop(ctx->au_);
    AudioUnitReset(ctx->au_, kAudioUnitScope_Global, 0);
    CloseComponent(ctx->au_);
    memset(&ctx->au_, 0, sizeof(ctx->au_));
    memset(&ctx->sd_, 0, sizeof(ctx->sd_));
  }
}

//----------------------------------------------------------------
// yae_au_ctx_render_cb
//
static OSStatus
callback(// The client data that is provided either
         // with the AURenderCallbackStruct
         // or as specified with the Add API call
         void * userData,

         // Flags used to describe more about the
         // context of this call (pre or post in the
         // notify case for instance)
         AudioUnitRenderActionFlags * ioActionFlags,

         // The times stamp associated with
         // this call of audio unit render
         const AudioTimeStamp * inTimeStamp,

         // The bus numbeer associated with this call
         // of audio unit render
         UInt32 inBusNumber,

         // The number of sample frames that will be
         // represented in the audio data in the
         // provided ioData parameter
         UInt32 inNumberFrames,

         // The AudioBufferList that will be used to
         // contain the rendered or provided audio data
         AudioBufferList * ioData)
{
  yae_au_ctx_t * ctx = (yae_au_ctx_t *)userData;
  assert(ioData->mNumberBuffers == 1);
  assert(ioData->mBuffers[0].mNumberChannels == ctx->sd_.mChannelsPerFrame);

  bool planar = !(ctx->sd_.mFormatFlags & kAudioFormatFlagsNativeFloatPacked);

  if (ctx->pull_(ctx->owner_,
                 ioData->mBuffers[0].mData,
                 inNumberFrames,
                 ctx->sd_.mChannelsPerFrame,
                 planar))
  {
    return noErr;
  }

  ctx->stop_(ctx->owner_);
  return noErr;
}

//----------------------------------------------------------------
// yae_au_ctx_open_stream
//
bool
yae_au_ctx_open_stream(void * context,
                       int sample_rate,
                       int num_channels)
{
  OSStatus err = noErr;
  yae_au_ctx_t * ctx = (yae_au_ctx_t *)context;

#if MAC_OS_X_VERSION_MAX_ALLOWED < 1060
  ComponentDescription desc;
#else
  AudioComponentDescription desc;
#endif

  memset(&desc, 0, sizeof(desc));
  desc.componentType = kAudioUnitType_Output;
  desc.componentSubType = kAudioUnitSubType_DefaultOutput;
  desc.componentManufacturer = kAudioUnitManufacturer_Apple;

#if MAC_OS_X_VERSION_MAX_ALLOWED < 1060
  Component comp = FindNextComponent(NULL, &desc);
#else
  AudioComponent comp = AudioComponentFindNext(NULL, &desc);
#endif

  memset(&ctx->au_, 0, sizeof(ctx->au_));

#if MAC_OS_X_VERSION_MAX_ALLOWED < 1060
  err = OpenAComponent(comp, &ctx->au_);
#else
  err = AudioComponentInstanceNew(comp, &ctx->au_);
#endif

  if (err)
  {
    return false;
  }

  UInt32 value = kAudioConverterQuality_Max;
  err = AudioUnitSetProperty(ctx->au_,
                             kAudioUnitProperty_RenderQuality,
                             kAudioUnitScope_Global,
                             0, // 0 == output
                             &value,
                             sizeof(value));
  if (err)
  {
    return false;
  }

  memset(&ctx->sd_, 0, sizeof(ctx->sd_));
  ctx->sd_.mFramesPerPacket  = 1;
  ctx->sd_.mBitsPerChannel   = sizeof(float) * 8;
  ctx->sd_.mFormatFlags      = kAudioFormatFlagsNativeFloatPacked;
  ctx->sd_.mFormatID         = kAudioFormatLinearPCM;
  ctx->sd_.mSampleRate       = sample_rate;
  ctx->sd_.mChannelsPerFrame = num_channels;
  ctx->sd_.mBytesPerPacket   = sizeof(float) * ctx->sd_.mChannelsPerFrame;
  ctx->sd_.mBytesPerFrame    = sizeof(float) * ctx->sd_.mChannelsPerFrame;

  err = AudioUnitSetProperty(ctx->au_,
                             kAudioUnitProperty_StreamFormat,
                             kAudioUnitScope_Input,
                             0, // 0 == output
                             &ctx->sd_,
                             sizeof(ctx->sd_));
  if (err)
  {
    AudioUnitReset(ctx->au_, kAudioUnitScope_Global, 0);
    CloseComponent(ctx->au_);
    memset(&ctx->au_, 0, sizeof(ctx->au_));
    return false;
  }

  AURenderCallbackStruct rcb;
  memset(&rcb, 0, sizeof(rcb));
  rcb.inputProc = &callback;
  rcb.inputProcRefCon = ctx;

  err = AudioUnitSetProperty(ctx->au_,
                             kAudioUnitProperty_SetRenderCallback,
                             kAudioUnitScope_Output,
                             0, // 0 == output
                             &rcb,
                             sizeof(rcb));
  if (err)
  {
    return false;
  }

  err = AudioUnitInitialize(ctx->au_);
  if (err)
  {
    AudioUnitReset(ctx->au_, kAudioUnitScope_Global, 0);

#if MAC_OS_X_VERSION_MAX_ALLOWED < 1060
    CloseComponent(ctx->au_);
#else
    AudioComponentInstanceDispose(ctx->au_);
#endif

    memset(&ctx->au_, 0, sizeof(ctx->au_));
    return false;
  }

  err = AudioOutputUnitStart(ctx->au_);
  if (err)
  {
    AudioUnitReset(ctx->au_, kAudioUnitScope_Global, 0);

#if MAC_OS_X_VERSION_MAX_ALLOWED < 1060
    CloseComponent(ctx->au_);
#else
    AudioComponentInstanceDispose(ctx->au_);
#endif

    memset(&ctx->au_, 0, sizeof(ctx->au_));
    return false;
  }

  return true;
}
